File: src/graphics/TArrow.js
/**
 * Created by simong on 2/21/17.
 */

import TLine from './TLine';
import TPoint from './TPoint';
import styles from './styles';
import util from './util';

export default class TArrow extends TLine {

    static ARROW_HEAD_SIZE = 12;
    static ARROW_OFFSET = 18;

    constructor(graphics, start, end, options) {
        super(graphics, start, end, options);

        this.mArrowOffset = (options.arrowOffset !== undefined) ? options.arrowOffset : TArrow.ARROW_OFFSET;
        this.mArrowHeadSize = TArrow.ARROW_HEAD_SIZE;
        this.mReverse = 1;
        this.mTrueLength = 0;

        this.mArrowHeadPath = this.mGraphics.addPath([[0, 0], [1, 0], [0,1]], {
            fillColor: options.fillColor || options.strokeColor || 'gray'
        });
        this.item.strokeColor = options.strokeColor || 'gray';
        this.mArrowHeadPath.closePath();
        this.mHeadPoint = [0, 0];

        this.mArrowHead = new TPoint(graphics, [0, 0], {size: 0});
        this.mArrowTail = new TPoint(graphics, [0, 0], {size: 0});

        this.update();
    }

    update() {
        if (!this.mArrowHeadPath || !this.mArrowHead || !this.mArrowTail) {
            return;
        }

        this.mTrueLength = util.distance(this.mStartPoint.x, this.mStartPoint.y, this.mEndPoint.x, this.mEndPoint.y);
        // Moved this to Load, OK?
        //this.mReverse = this.mTrueLength < this.mArrowOffset ? -1 : (this.mTrueLength == 0 ? 0 : 1);

        let tempDir = util.direction(this.mEndPoint.x, this.mEndPoint.y, this.mStartPoint.x, this.mStartPoint.y);

        this.mHeadPoint = [(this.mEndPoint.x + this.mArrowOffset * Math.cos(tempDir)),
            (this.mEndPoint.y + this.mArrowOffset * Math.sin(tempDir))];

        const arrowPoint = (this.mReverse > 0) ? this.mHeadPoint : [this.mStartPoint.x, this.mStartPoint.y];
        this.mArrowHead.item.position = arrowPoint;
        //this.mArrowHead.item.position = (this.mReverse < 0) ? [this.mStartPoint.x, this.mStartPoint.y] : this.mHeadPoint;
        this.mArrowTail.item.position = (this.mReverse < 0) ? this.mHeadPoint : [this.mStartPoint.x, this.mStartPoint.y];

        this.mArrowHeadPath.segments[0].point.x = (arrowPoint[0] - (this.mReverse * this.mSize) * Math.cos(tempDir));
        this.mArrowHeadPath.segments[0].point.y = (arrowPoint[1] - (this.mReverse * this.mSize) * Math.sin(tempDir));
        tempDir += 15.0;
        this.mArrowHeadPath.segments[1].point.x = (arrowPoint[0] - (this.mReverse * this.mArrowHeadSize) * Math.cos(tempDir));
        this.mArrowHeadPath.segments[1].point.y = (arrowPoint[1] - (this.mReverse * this.mArrowHeadSize) * Math.sin(tempDir));
        tempDir -= 30.0;
        this.mArrowHeadPath.segments[2].point.x = (arrowPoint[0] - (this.mReverse * this.mArrowHeadSize) * Math.cos(tempDir));
        this.mArrowHeadPath.segments[2].point.y = (arrowPoint[1] - (this.mReverse * this.mArrowHeadSize) * Math.sin(tempDir));
        super.update();

        if (this.mLabel) {
            this.mLabel.content = this.mLabelText || '';
            this.mLabel.position = [this.mLabelOffset[0] + (this.mStartPoint.x + this.mHeadPoint[0]) / 2.0,
                this.mLabelOffset[1] + (this.mStartPoint.y + this.mHeadPoint[1]) / 2.0];
        }
    }

    length() {
        if (!this.mHeadPoint) {
            return 0;
        }
        return util.distance(this.mStartPoint.x, this.mStartPoint.y, this.mHeadPoint[0], this.mHeadPoint[1]);
    }

    moment(pivot) {
        return TLine.ccw(this.mStartPoint.x, this.mStartPoint.y, this.mHeadPoint[0], this.mHeadPoint[1], pivot.x, pivot.y) * this.length() * this.perpDist(pivot.x, pivot.y);
    }
};

File: old_java_src/ActiveStatics/src/truss/TArrow.java
package truss;

/**
 * Title:        Truss
 * Description:
 * Copyright:    Copyright (c) 2001
 * Company:
 * @author Simon Greenwold
 * @version 1.0
 */
import java.awt.*;

public class TArrow extends TLine {

  public int ARROW_OFFSET = 15;
  public static final float ARROW_HEAD_SPREAD = 15; //(float)(15 * Math.PI / 180.0);
  public static final int ARROW_HEAD_SIZE = 12;
  public int mArrowOffset;
  public float mArrowHeadSpread;
  public int mArrowHeadSize;
  public TPoint mArrowHead;
  int xHeadPoints[] = new int[3];
  int yHeadPoints[] = new int[3];
  int mReverse = 1;
  float mTrueLength;

  protected float tempDir;

  public TArrow() {
    mArrowOffset = ARROW_OFFSET;
    mArrowHeadSpread = ARROW_HEAD_SPREAD;
    mArrowHeadSize = ARROW_HEAD_SIZE;
    mArrowHead = new TPoint();
  }

  public void update() {
    mTrueLength = Util.distance(mStartPoint.x, mStartPoint.y, mEndPoint.x, mEndPoint.y);
    if (mTrueLength < mArrowOffset) {
      mReverse = -1;
    }
    else if (mTrueLength == 0) {
      mReverse = 0;
    }
    else
      mReverse = 1;

    tempDir = Util.direction(mEndPoint.x, mEndPoint.y, mStartPoint.x, mStartPoint.y);
    mArrowHead.x = (float)(mEndPoint.x + mArrowOffset * Math.cos(tempDir));
    mArrowHead.y = (float)(mEndPoint.y + mArrowOffset * Math.sin(tempDir));

    xHeadPoints[1] = (int)(mArrowHead.x - (mReverse * mSize) * Math.cos(tempDir));
    yHeadPoints[1] = (int)(mArrowHead.y - (mReverse * mSize) * Math.sin(tempDir));

    tempDir += ARROW_HEAD_SPREAD;

    xHeadPoints[0] = (int)(mArrowHead.x - mReverse * mArrowHeadSize * Math.cos(tempDir));
    yHeadPoints[0] = (int)(mArrowHead.y - mReverse * mArrowHeadSize * Math.sin(tempDir));

    tempDir -= 2 * ARROW_HEAD_SPREAD;

    xHeadPoints[2] = (int)(mArrowHead.x - mReverse * mArrowHeadSize * Math.cos(tempDir));
    yHeadPoints[2] = (int)(mArrowHead.y - mReverse * mArrowHeadSize * Math.sin(tempDir));
  }

  public void drawDashedLine(Graphics g, int x1, int y1, int x2, int y2, int thickness) {
   float fromX, toX, fromY, toY;
   float dist = Util.distance(x1, y1, x2, y2);
   float dir = Util.direction(x1, y1, x2, y2);
   float fromDist = 0.0f;
   float toDist = 0.0f;
   float cos = (float)Math.cos(dir);
   float sin = (float)Math.sin(dir);

   while (fromDist < dist) {
     toDist = (float)Math.min(fromDist + mDashLength, dist);
     fromX = x1 + fromDist * cos;
     fromY = y1 + fromDist * sin;
     toX = x1 + toDist * cos;
     toY = y1 + toDist * sin;

     drawThickLine(g, (int)fromX, (int)fromY, (int)toX, (int)toY, thickness);

     fromDist = toDist + mGapLength;
   }

  }


  public void draw(Graphics graphics) {
    if (mInvisible)
      return;

    graphics.setColor(mColor);
    if (!mDashed) {
      drawThickLine(graphics, (int) mStartPoint.x, (int) mStartPoint.y,
                    (int) mArrowHead.x, (int) mArrowHead.y, mSize);
    }
    else {
      drawDashedLine(graphics, (int) mStartPoint.x, (int) mStartPoint.y,
                    (int) mArrowHead.x, (int) mArrowHead.y, mSize);
    }
    graphics.fillPolygon(xHeadPoints, yHeadPoints, 3);
    if (!mLabel.equals("")) {
      graphics.setFont(G.mLabelFont);
      graphics.setColor(Color.black);
      graphics.drawString(mLabel, mLabelXOff + (int)((mStartPoint.x + mArrowHead.x) / 2.0f),
                                  mLabelYOff + (int)((mStartPoint.y + mArrowHead.y) / 2.0f));
    }
  }

  public boolean hit(Point p) {
    return false;
  }

  public float length() {
    return Util.distance(mStartPoint.x, mStartPoint.y, mArrowHead.x, mArrowHead.y);
  }

  public float moment(TPoint pivot) {
    return  CCW(mStartPoint.x, mStartPoint.y, mArrowHead.x, mArrowHead.y, pivot.x, pivot.y) *
            length() *
            perpDist(pivot);
  }
}